src/middlewares/createTelemetryMiddleware.ts

```diff
@@ -0,0 +1,110 @@
+import { InstantSearch, Middleware, Widget } from '../types';
+import { Index } from '../widgets/index/index';
+
+type TelemetryWidget = {
+  type: string;
+  params: string[];
+  officialWidget: boolean;
+};
+
+type TelemetryPayload = {
+  widgets: TelemetryWidget[];
+};
+
+function extractPayload(
+  widgets: Array<Widget<{ renderState: any }>>,
+  instantSearchInstance: InstantSearch,
+  payload: TelemetryPayload
+) {
+  const parent = instantSearchInstance.mainIndex;
+
+  const initOptions = {
+    instantSearchInstance,
+    parent,
+    scopedResults: parent.getScopedResults(),
+    state: parent.getHelper()!.state,
+    helper: parent.getHelper()!,
+    createURL: parent.createURL,
+    uiState: instantSearchInstance._initialUiState,
+    renderState: instantSearchInstance.renderState,
+    templatesConfig: instantSearchInstance.templatesConfig,
+    searchMetadata: {
+      isSearchStalled: instantSearchInstance._isSearchStalled,
+    },
+  };
+
+  widgets.forEach(widget => {
+    let widgetParams = {};
+
+    if (widget.getWidgetRenderState) {
+      const renderState = widget.getWidgetRenderState(initOptions);
```

Comment:
I actually had a look, and while we are missing the parameters that are not sent to the connector (visual-only ones), we still retrieve all the values without their defaults, since those are only assigned on the destructuring on the connector. I think that's a fair tradeoff that doesn't require rewriting all widgets. I will however after this do a pass on all widgets, and make sure they don't set any defaults that are also set in the connectors

Reply:
I think in a second pass, we actually could send the visual-only parameters to the connector as well, since it's impossible for the parameters to clash significantly

Reply:
I think this is a mandatory step, in the current PR we are missing a lot of information (see https://github.com/algolia/instantsearch.js/blob/2431b966e771d8d9386d919eb469c79b27b5b24a/src/widgets/pagination/pagination.js for example, only `totalPages` and `padding` are sent to the connector, we are missing `templates`, `showFirst`, `showLast`, etc.)

Also in most widget the parameters are send after destructuring, so with their default values.

Reply:
> so with their default values

I've checked, and that's not really true. They're destructured, but the default value is applied in the connector